### PR TITLE
[Review] Introduce Deterministic Factory

### PR DESCRIPTION
**Pull request by marvinkruse**
Just removed and re-added the contracts so we have a PR that can be reviewed. Won't be merged.

### CODE REVIEW

File: script/Deployment.s.sol [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    /// @return deterministicFactory The address of the deployed Deterministic Factory
+    function run() public virtual returns (address deterministicFactory) {
+        console2.log(
+            "-----------------------------------------------------------------------------"
+        );
+        console2.log("Deploy Inverter Deterministic Factory...");
+
+        if (multisig == address(0)) {
+            console2.log("No MULTISIG_ADDRESS set, using deployer as owner.");
+            multisig = deployer;
```
Comment (Zitzak) [inline script/Deployment.s.sol:26]: Is this set like this for testing purposes I assume? If not I would just let it revert to ensure the multisig has been set

Comment (Zitzak) [reply script/Deployment.s.sol:26]: Within `PrintBytecode` script you did exactly that

File: src/DeterministicFactory_v1.sol [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+// External Dependencies
+import {Ownable2Step, Ownable} from "@oz/access/Ownable2Step.sol";
+import {Create2} from "@oz/utils/Create2.sol";
+
+/**
+ * @title   Deterministic Deployment Factory
+ *
+ * @notice  This contract is a factory for deploying contracts using the CREATE2 opcode,
+ *             allowing us to deploy some of our contract at the same address of each EVM-
+ *          compatible network.
```
Comment (Zitzak) [inline src/DeterministicFactory_v1.sol]:
```suggestion
 * @notice  This contract is a factory for deploying contracts using the CREATE2 opcode,
 *              allowing us to deploy some of our contract at the same address for each EVM-
 *              compatible network.
```

Comment (Zitzak) [reply src/DeterministicFactory_v1.sol]: Formatting off and changed one word

Comment (Zitzak) [reply src/DeterministicFactory_v1.sol]: We also might want to add that the deterministic deployment does not work for ZK EVM chains, as spoken about during our call

File: src/DeterministicFactory_v1.sol [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * @dev     We limit this to a specific address ('allowedDeployer'), as otherwise anyone
+ *          could deploy our contracts at the addresses we use. This can be problematic
+ *          for contracts that are initialized, i.e. where the owner is only set
+ *          after deployment, which would hand them ownership of the contracts. We intend
+ *          to set this 'allowedDeployer' to the deployment account we use, and then
+ *          set it to the zero address afterwards (until the next deployment takes place).
```
Comment (Zitzak) [inline src/DeterministicFactory_v1.sol]:
```suggestion
 * @dev     We limit the deployment to a specific address ('allowedDeployer'), as otherwise anyone
 *          could deploy our contracts at the addresses we use. This can be problematic
 *          for contracts that are initialized, i.e. where the owner is only set
 *          after deployment, which would hand them ownership of the contracts. We intend
 *          to set this 'allowedDeployer' to the deployment account we use, and then
 *          set it to the zero address afterwards (until the next deployment takes place).
```

Comment (Zitzak) [reply src/DeterministicFactory_v1.sol]: Given what has been written here, do we need a script for setting/removing the `allowedDeployer` or will this be done in the UI? I can imagine we need a script for the bytecode to use in a custom Tx? Also this should be in the guide you are wrting such that everyone of the founders can do it.

File: src/DeterministicFactory_v1.sol [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+ *                          In case of any concerns or findings, please refer to our Security Policy
+ *                          at security.inverter.network or email us directly!
+ *
+ * @author  Inverter Network
+ */
+contract DeterministicFactory_v1 is IDeterministicFactory_v1, Ownable2Step {
+    //--------------------------------------------------------------------------
+    // Variables
+
+    // The address that is allowed to use the factory.
```
Comment (Zitzak) [inline src/DeterministicFactory_v1.sol]:
```suggestion
    /// @dev The address that is allowed to use the factory.
```

Comment (Zitzak) [reply src/DeterministicFactory_v1.sol]: For the automated natspec detection to work.

File: src/DeterministicFactory_v1.sol [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    // Public Functions
+
+    /// @inheritdoc IDeterministicFactory_v1
+    function setAllowedDeployer(address _allowedDeployer) external onlyOwner {
+        allowedDeployer = _allowedDeployer;
+        emit DeterministicFactory__AllowedDeployerChanged(_allowedDeployer);
+    }
+
+    /// @inheritdoc IDeterministicFactory_v1
+    function deployWithCreate2(bytes32 salt, bytes calldata code)
```
Comment (Zitzak) [inline src/DeterministicFactory_v1.sol:66]: As we pass the salt into the function, where/how is it created? I was assuming that it would be generated from the msg.sender PK somehow such that it can't be replicated. Wouldn't it in this way be possible to create their own factory, find the parameters on a block explorer and deploy the contracts? Or am I missing something?

Comment (Zitzak) [reply src/DeterministicFactory_v1.sol:66]: Nvm, indeed I missed something here. It automatically uses the senders address under the hood. Had to look it up.

One question for me remains still. Where and how do define the salt. Wouldn't it be easiest to just have it in the factory itself?

**Review by Zitzak — CHANGES_REQUESTED** [review]
Few things I've found. One thing I would like to note here, which we all should become strict about after mainnet launch, is keeping to the [Inverter Testing Handbook](https://docs.google.com/document/d/1YEf7JoscVhcwTmJst-twoYSUFBQ5O_IrX4uQwhLLAPA/edit#heading=h.81p2puwq7pfd), e.g. the way the test functions are call, writing test tree etc. Didn't comment withint the tests for now due to time constraints

File: script/Deployment.s.sol (hunk 1 above)
Comment (marvinkruse) [reply script/Deployment.s.sol:26]: Yes, I did that because the Contract will not be deployed via Foundry on live networks, only for testing. The way it will be deployed is via `PrintBytecode` and then as a tx in safe Safe UI. That's why I don't revert in the `Deployment` script, as that's just for testing.

File: src/DeterministicFactory_v1.sol (hunk 2 above)
Comment (marvinkruse) [reply src/DeterministicFactory_v1.sol]: It does work for ZK EVM chains, so I won't add anything here. The thing is just that zk EVMs and EVMs don't share the same VM bytecode, and deterministic math operations will only result in the same address if the network uses the same VM bytecode as the other chains you are on. So for all regular EVM chains, our address will be the same (e.g. `0x123...123` and for all ZK EVM chains the address will be the same (e.g. `0.333...333`). 

We'd still use it on ZK EVM chains, as a lot of them share the same VM bytecode, but otherwise, that's just an inherent nature of how CREATE2 works, so I didn't add it here. 

I don't really see the value in adding it here, as it's nothing specific to our factory or approach. But happy to add that if you still think it should be mentioned?

File: src/DeterministicFactory_v1.sol (hunk 3 above)
Comment (marvinkruse) [reply src/DeterministicFactory_v1.sol]: This will be done in the Safe UI. As the contract will be verified, we don't need a script for this for mainnets. I can still add one for testing purposes though, it's a good idea.

(And yes, that is part of the guide document)

File: src/DeterministicFactory_v1.sol (hunk 5 above)
Comment (marvinkruse) [reply src/DeterministicFactory_v1.sol:66]: > It automatically uses the senders address under the hood. Had to look it up.

It does **not** use the senders address under the hood, that is the whole idea of using CREATE2. The input factory for CREATE2 are the address of the executing entity (which in this case is the factory contract), as well as the bytecode of the created contract (including constructor data) and the salt.

> Where and how do define the salt. Wouldn't it be easiest to just have it in the factory itself?

The salt is defined outside of the factory, like explained in the `USAGE.md` file, that can look like this: 
```solidity
bytes32 salt = keccak256("inverter-deployment-01");
bytes memory implBytecode = vm.getCode("Governor_v1.sol:Governor_v1");
address govImplementation = factory.deployWithCreate2(salt, implBytecode);
```
The using entity (in our case we use it for the protocol) then needs to handle the use of the right salt at the right time (sometimes different salts may be used during deployment, if otherwise same-bytecode-contracts can't be deployed due to address collisions). So it can't be defined by the factory itself. (I already spoke about this with Felix on how we'll do that in the protocol repo)

Hope that clears it up?

Comment (Zitzak) [reply src/DeterministicFactory_v1.sol:66]: Okay then I mistook the sender described in the OZ docs as meaning the msg.sender.

**Comment by Zitzak** [comment]
I forgot one more thing to mention in the initial review.

Could you please attach logs/printscreen from your manual deployment test, as we have spoken about within our dev sync. Thank you @marvinkruse

**Comment by marvinkruse** [comment]
@Zitzak 
I improved the tests again and added trees for each of the states/behaviors as well in [this commit](https://github.com/InverterNetwork/deterministic-factory/pull/1/commits/304e96aa78bea33d32ca363993532b89fd6f9907), making sure that the tests all all covering these at least once. Do you think it's better? Thank you for the nudge to improve here and follow our standards :)

Coverage is also fine (_was before already, was just making sure again_).

Screenshot 2024-08-01 at 22 31 30


> Could you please attach logs/printscreen from your manual deployment test, as we have spoken about within our dev sync.

Yes, will do tomorrow 👍  _(Edit: Done, see next comment.)_

**Comment by marvinkruse** [comment]
@Zitzak @0xNuggan @FHieser 
I did a documented test run where I:
* Deployed a multisig on different networks (at different nonces, but yielding same address due to create2)
* Deployed the DeterministicFactory via these Multisigs via the CreateCall library at the same address
* Deployed a contract on the two networks, at the same address, even when using different deployer EOAs with different nonces

So basically all we need for our deployment. All documented [here](https://docs.google.com/document/d/1KXzc7IM2YVcSX9WHEEApHKpXt0PKWwliNNEUyV_5kxo).

**Review by 0xNuggan — APPROVED** [review]
